### ISSUE TITLE
BCEWithLogitsLoss() not equal to BCELoss() with sigmoid()

### ISSUE THREAD

**Issue by ketanhdoshi** [pytorch/pytorch#39435]
Calling BCEWithLogitsLoss() gives different results than calling sigmoid with BCELoss(). Please see the example below.

pred = tensor(
       [[ 1.6094e+01],
        [ 5.1216e-01],
        [-9.5367e+00],
        [ 2.4377e+01],
        [ 3.5349e+01],
        [ 2.5894e+01],
        [ 3.1808e+01],
        [ 1.8592e-01],
        [-2.8377e+00],
        [-9.0343e+00],
        [ 1.5208e+01],
        [ 4.8932e+00],
        [-5.4581e+00],
        [ 1.8190e+01],
        [ 1.0950e+01],
        [ 1.7572e+01],
        [-2.2359e+01],
        [ 2.0425e+01],
        [ 1.9340e+02],
        [ 8.5044e-01]])

targ = tensor([[0.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.]])

print (F.binary_cross_entropy_with_logits(pred, targ), nn.BCEWithLogitsLoss()(pred, targ), nn.BCELoss()(torch.sigmoid(pred), targ))

# Error still occurs with float64
pred, targ = pred.double(), targ.double()
print (F.binary_cross_entropy_with_logits(pred, targ), nn.BCEWithLogitsLoss()(pred, targ), nn.BCELoss()(torch.sigmoid(pred), targ))

# Error still occurs with two samples
# NB: However the error goes away with two samples with float64
pred, targ = pred[:2], targ[:2]
print (F.binary_cross_entropy_with_logits(pred, targ), nn.BCEWithLogitsLoss()(pred, targ), nn.BCELoss()(torch.sigmoid(pred), targ))

# Try with these values instead and the error gets magnified even more
pred2 = tensor(
       [[ 1.6094e+01],
        [ 5.1216e-01],
        [-9.5367e+00],
        [ 2.4377e+01],
        [ 3.5349e+01],
        [ 2.5894e+12],
        [ 3.1808e+01],
        [ 1.8592e-01],
        [-2.8377e+00],
        [-9.0343e+00],
        [ 1.5208e+01],
        [ 4.8932e+00],
        [-5.4581e+00],
        [ 1.8190e+01],
        [ 1.0950e+07],
        [ 1.7572e+06],
        [-2.2359e+01],
        [ 2.0425e+01],
        [ 1.9340e+02],
        [ 8.5044e-01]])

Pytorch - 1.5.0+cu101
Python - 3.6.9
Results are the same with CPU and GPU

**Comment by ptrblck** [comment]
`nn.BCEWithLogitsLoss` gives you a higher numerical stability as described in https://github.com/pytorch/pytorch/issues/751.

Have a look at [this code snippet](https://discuss.pytorch.org/t/difference-between-bceloss-and-bcewithlogitsloss-when-using-large-values/54053/2?u=ptrblck) to check the results manually.

**Comment by ketanhdoshi** [comment]
Thanks for the code snippet, @ptrblck. Based on that, is it fair to conclude that BCELoss does not provide accurate loss for preds > 10 or so?

If so, it might make sense to mention this in the documentation, or provide a warning message?

**Comment by ptrblck** [comment]
I wouldn't necessarily specify a certain threshold, as it also depends on the data type you are using.
`nn.BCEWithLogitsLoss` contains this information:
> This loss combines a Sigmoid layer and the BCELoss in one single class. This version is more numerically stable than using a plain Sigmoid followed by a BCELoss as, by combining the operations into one layer, we take advantage of the log-sum-exp trick for numerical stability.

but if you feel it's not clear, we would be happy to get a PR for an improved documentation. :)
